### PR TITLE
BUG: args offset in Timestamp.__new__ causes bugs

### LINKED ISSUE #52343: BUG: args offset in Timestamp.__new__ causes bugs
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
# Pandas 1.5.x
import pandas as pd

ts1 = pd.Timestamp(2023, 3, 30, 10, None, None, 30)
ts2 = pd.Timestamp(2023, 3, 30, 10, day=30)

> 2023-03-30 10:00:00.000030
> 2023-03-30 10:00:00.000030

# Pandas 2.x
ts1 = pd.Timestamp(2023, 3, 30, 10, None, 30)
ts2 = pd.Timestamp(2023, 3, 30, 10, second=30)

> 2023-03-30 10:00:00.000030
> 2023-03-30 10:00:00.000030
```

### Issue Description

When we build `Timestamp` from components through positional args, the year initially ends up in `ts_input`, and we need to shift args one to the right, which we do with some kludges (normally by calling a constructor and passing it arguments in offset positions).

The problem is that we cannot tell the difference between the construction calls given in the example from inside the current `Timestamp.__new__`. This is the cause of both the bug outlined here and also of #52117 .

The two available solutions are  to either decorate `Timestamp`, which causes a performance hit (see #52221) or to refactor `Timestamp` to be built from `*args, **kwargs` and handle the logic internally.

### Expected Behavior
```python
# Pandas 2.x
ts1 = pd.Timestamp(2023, 3, 30, 10, None, None, 30)
ts2 = pd.Timestamp(2023, 3, 30, 10, second=30)

> 2023-03-30 10:00:00.000030
> 2023-03-30 10:00:30
```
### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 8c7b8a4f3eaf4aef90c300baa9a2337323f7a12b
python           : 3.8.16.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.22621
machine          : AMD64
processor        : AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1251

pandas           : 2.1.0.dev0+337.g8c7b8a4f3e
numpy            : 1.23.5
pytz             : 2023.2
dateutil         : 2.8.2
setuptools       : 67.6.0
pip              : 23.0.1
Cython           : 0.29.33
pytest           : 7.2.2
hypothesis       : 6.70.0
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.9
lxml.etree       : 4.9.2
html5lib         : 1.1
pymysql          : 1.0.2
psycopg2         : 2.9.3
jinja2           : 3.1.2
IPython          : 8.11.0
pandas_datareader: None
bs4              : 4.12.0
bottleneck       : 1.3.7
brotli           : 
fastparquet      : 2023.2.0
fsspec           : 2023.3.0
gcsfs            : 2023.3.0
matplotlib       : 3.6.3
numba            : 0.56.4
numexpr          : 2.8.3
odfpy            : None
openpyxl         : 3.1.0
pandas_gbq       : None
pyarrow          : 11.0.0
pyreadstat       : 1.2.1
pyxlsb           : 1.0.10
s3fs             : 2023.3.0
scipy            : 1.10.1
snappy           : 
sqlalchemy       : 2.0.7
tables           : 3.7.0
tabulate         : 0.9.0
xarray           : 2023.1.0
xlrd             : 2.0.1
zstandard        : 0.19.0
tzdata           : 2023.2
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by AlexKirko**
- [x] closes #52343 , #52117 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] ~~Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.~~
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This is the promised refactor for the Timestamp constructor. Detailed comments are below.

ASV benchmark results for Timestamp construction:
```
       before           after         ratio
     [1c2ad16b]       [7378f54f]
     <main>           <gh-52343-timestamp-from-positional>
+        1.20±0μs      1.86±0.02μs     1.56  tslibs.timestamp.TimestampConstruction.time_from_positional
-     1.05±0.01μs         953±10ns     0.91  tslibs.timestamp.TimestampConstruction.time_parse_iso8601_no_tz
-         253±1ns          209±4ns     0.82  tslibs.timestamp.TimestampConstruction.time_from_pd_timestamp
```
Replaces #52221

### CODE REVIEW

**Comment by AlexKirko** [comment]
WIP. The same tests as in #52221, different solution.

File: asv_bench/benchmarks/tslibs/timestamp.py [hunk 1]
```diff
@@ -48,6 +48,9 @@ def time_from_datetime_aware(self):
     def time_from_pd_timestamp(self):
         Timestamp(self.ts)
 
+    def time_from_positional(self):
+        Timestamp(2020, 1, 1, 0, 0, 0)
```
Comment (AlexKirko) [inline asv_bench/benchmarks/tslibs/timestamp.py:52]: An additional benchmark for future regressions might be helpful.

**Comment by AlexKirko** [comment]
@mroeschke @jbrockmendel @jreback 
Hello, everyone. Here is the Timestamp refactor to using `*args, **kwargs` that I promised. This version fixes both the fold and the positional vs keyword inconsistency bugs. There is some performance gain when building from a Timestamp. There is a performance loss when building from positional arguments (added a check to the asv benchmark suite).

Had to rewrite most of the constructor, because naively switching to `*args, **kwargs` without changing the rest of the code had resulted in a 50-60% hit for most of the use cases.

Please review. Any suggestions on improving the performance of the positional arguments case would be appreciated.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 2]
```diff
@@ -1556,17 +1539,76 @@ class Timestamp(_Timestamp):
...
-        _date_attributes = [year, month, day, hour, minute, second,
-                            microsecond, nanosecond]
+        args_len = len(args)
+
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        # checking verbosely, because cython doesn't optimize
+        # list comprehensions (as of cython 0.29.x)
+        if args_len == 1 and len(kwargs) == 0 and isinstance(args[0], _Timestamp):
+            return args[0]
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timestamps.pyx:1557]: What if a user passes `Timestamp(Timestamp(...), tz="UTC")`? I _think_ that should localize the timestamp?

File: pandas/_libs/tslibs/timestamps.pyx [hunk 3]
```diff
@@ -1556,17 +1539,76 @@ class Timestamp(_Timestamp):
...
+
+        # Building from ts_input
+        if args_len == 1:
+            if kwargs:
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timestamps.pyx:1561]: Should we raise if a user passes `Timestamp("2020-01-01" , foo="bar")`?

File: pandas/_libs/tslibs/timestamps.pyx [hunk 4]
```diff
@@ -1556,17 +1539,76 @@ class Timestamp(_Timestamp):
...
+            ts_input = args[0]
+            tzinfo = kwargs.get("tzinfo")
+        # Building from positional arguments
+        elif 9 > args_len > 2 and isinstance(args[1], int):
+            args = args + (None,) * (8 - args_len)
+            year, month, day, hour, minute, second, microsecond, tzinfo = args
+
+            if kwargs:
+                # Positional or keyword arguments
+                hour = kwargs.get("hour", hour)
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timestamps.pyx:1581]: Shouldn't we raise if a user passes the same argument by position and keyword e.g.

```
In [4]: datetime.datetime(2020, 1, 1, day=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 datetime.datetime(2020, 1, 1, day=1)

TypeError: argument for function given by name ('day') and position (3)
```

**Comment by AlexKirko** [comment]
Doesn't look like the Parquet arrows in CircleCI have anything to do with this PR.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 5]
```diff
@@ -1556,17 +1539,95 @@ class Timestamp(_Timestamp):
...
+                minute = kwargs.get("minute", minute)
+                second = kwargs.get("second", second)
+                microsecond = kwargs.get("microsecond", microsecond)
+                tzinfo = kwargs.get("tzinfo", tzinfo)
+        # Keywords only
+        elif args_len == 0:
+            ts_input = kwargs.get("ts_input", _no_input)
+            year = kwargs.get("year")
+            month = kwargs.get("month")
+            day = kwargs.get("day")
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1610]: is there anyway to simply translate the args to kwargs near the top

this looks so fragile and error prone

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
i know i suggested the args/kwargs pattern, but... is there an alternative way out of the current problem by being stricter in disallowing mixed positional/keyword?

**Comment by jreback** [comment]
> i know i suggested the args/kwargs pattern, but... is there an alternative way out of the current problem by being stricter in disallowing mixed positional/keyword?

+1 in this (or allow only fully qualified positionals - eg ymdhms required plus kwarg for example)

**Comment by AlexKirko** [comment]
@jbrockmendel @jreback  Another option is to just use the Python>3.8 [args-only syntax](https://peps.python.org/pep-0570/) to make everything up to tzinfo positional-only. A couple of issues I see:

* this breaks the behavior compatibility with datetime. Users will construct the way they do with `datetime` and get errors. WIll need to make those clear enough that they don't think this is unexpected behavior.
* this breaks old code. I personally built Timestamps like `Timestamp(2020, 1, 1, hour=20, minute=30, second=0)` in my applications, adding kwargs only where I needed to make it clear what the Timestamp would contain. This is a pretty standard practice in finance where I come from.
* going through a deprecation cycle would keep the bug unfixed (unless we temporarily go with decoration or the current approach and add deprecation warnings, which is what we should do in that case)
* The arg offset bug is present in version 1.5.x too. A fix that breaks old code would not be possible/available there.
* tzinfo would become positional-only, I assume. tz will remain kwarg-only. This would look weird to users and probably be reported as an issue.

Because of these points, I didn't think this option was on the table.

Let's come up with an approach that you will be able to accept if it's decently done, and I will implement it. But please be aware that I won't be redoing this from scratch a fourth time, so let's get it right (I'll be happy to go back to one of the previous options though if we don't like the third solution).

I believe that the current arg offset in Timestamp is a buggy PITA that we took advantage of to maintain performance and keep code readable. The fix will either be a kludge, be harder to maintain, or break a bunch of code. I don't see a solution that does neither of those but would be happy to be wrong.

**Comment by AlexKirko** [comment]
Unfortunately, I don't think there is a way of fixing this outside of using `*args, **kwargs` or enforcing the positional/kwargs division.

**Comment by jbrockmendel** [comment]
@AlexKirko you make goods points, thanks for thinking through the idea that i clearly didn't.  I'll give this a close look.  Thanks for being patient.

File: pandas/_libs/tslibs/timestamps.pyx [hunk 6]
```diff
@@ -1523,31 +1523,14 @@ class Timestamp(_Timestamp):
...
-        second=None,
-        microsecond=None,
-        tzinfo_type tzinfo=None,
-        *,
-        nanosecond=None,
-        tz=None,
-        unit=None,
-        fold=None,
-    ):
+    def __new__(cls, *args, **kwargs):
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timestamps.pyx:1526]: does this get rendered in the docs somewhere?

**Comment by AlexKirko** [comment]
A bit sick at the moment, will come back to it in a couple days once I'm better.

**Comment by AlexKirko** [comment]
Back among the living. The illness and the job change have cost me some time.

File: doc/source/whatsnew/v2.1.0.rst [hunk 7]
```diff
@@ -368,6 +368,8 @@ Datetimelike
 - Bug in :meth:`Timestamp.round` with values close to the implementation bounds returning incorrect results instead of raising ``OutOfBoundsDatetime`` (:issue:`51494`)
 - Bug in :meth:`arrays.DatetimeArray.map` and :meth:`DatetimeIndex.map`, where the supplied callable operated array-wise instead of element-wise (:issue:`51977`)
 - Bug in constructing a :class:`Series` or :class:`DataFrame` from a datetime or timedelta scalar always inferring nanosecond resolution instead of inferring from the input (:issue:`52212`)
+- Bug in :class:`Timestamp` raising an error when passing fold when constructing from positional arguments.
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.1.0.rst:371]: can you put "fold" in quotes/ticks

File: pandas/_libs/tslibs/timestamps.pyx (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx:1755]: @AlexKirko any idea about this?

**Comment by jbrockmendel** [comment]
Is there any viable path to splitting this into smaller pieces?  This seems like a very difficult problem and is easy for me and other reviewers to be lazy about reviewing

**Comment by AlexKirko** [comment]
> Is there any viable path to splitting this into smaller pieces? This seems like a very difficult problem and is easy for me and other reviewers to be lazy about reviewing

The bigger problem, to be honest, is me having been either swamped with work or sick this past month.

In my opinion, splitting this into multiple PRs that make sense is possible, only if we stop caring about the performance of the initial implementation, and put the necessary performance improvements into separate PRs. It would make this easier to review but lead to an initial PR that is very slow in some use cases.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by AlexKirko** [comment]
I have merged in main, and I will try another approach next week, so I would like to reopen the request for now.

The problem here is that it's a "I will hate it when I see it" situation. There are three solutions that come to mind, and I've tried and submitted for review two. The third is terrible, in my opinion.
1. decorate the constructor. Very slow, very easy to read and support.
2. fix the kwargs/args order inside the constructor. Minimal performance impact, but is ugly and makes the code harder to support. This is the current approach in the branch that everyone seems to also reject.
3. Use the positional-only syntax. Breaks compatibility with datetime, breaks user code. I think it's much worse than the previous two approaches, but I can try it next week for illustrative purposes if nothing else.

No one has suggested another solution, and I'm out of ideas. I personally think that option 2 (the current one) is fine. Yes, it's ugly, but it doesn't slow things down too much or break old code and datetime compatibilty.

I'm well aware that all the approaches have significant downsides, but either someone needs to suggest an idea I haven't considered, or we need to choose. I'll try out the positional-only solution next week for illustration purposes, but I think it's the worst. Breaking API compatibility with current code to fix a bug is just wrong.

We can also leave the bug in if fixing it is worse than the current state of things, but that's also not great, to put it in mildly.

If I had the authority to take the responsibility for this decision, I'd just pick the current solution and move on.

I'd be happy to hear your thoughts, @mroeschke . You have more experience, so maybe you see a way out here.

**Review by jreback — COMMENTED** [review]
option 2 (current approach) looks like a good soln

the code complexity is not too bad - commented pretty well already

**Comment by AlexKirko** [comment]
Okay, then I'll focus on cleaning up the current solution instead. Will try to make the changes as painless as possible to support.

**Comment by AlexKirko** [comment]
I apologize for being glacially slow on this. Finally had an opportunity to review.

The MR is in decent shape except for the docs issue. We change the signature to `*args, **kwargs` and this gets reflected in the auto-generated docs. I suppose there is nothing to do here except implement an ad-hoc override for the docs.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**